Correct out-of-spec Avro schemas with typing.Union

xgamer4 opened the pull request:
This pull request corrects  #85 by checking the type of the provided default and placing it at the front of the type field in Avro schema. It adds two tests - one to reproduce the problem detailed in Issue 85 to develop against, and one to add a test to explicitly document the behavior of a `typing.Optional[T]` field without a default. The behavior of a `typing.Optional[T]` field remains unchanged. The existing test for union fields with defaults was updated. 

Documentation still needs to be updated, and notes on the change (it's likely a breaking change, and notes on the severity) need to be added. 

I wouldn't be surprised if there were minor merge conflicts with #86 as well - the two branches were developed independently of each other - so I'll need to reconcile those if/when one of these two merge requests gets approved.

xgamer4 commented:
Testing against the released v0.19.0 for breaking changes, I built a class like this:
```python
@dataclass
class RandomSentence(faust.Record, AvroModel, serializer='avro_randomsentence'):
    """Randomly Generated Sentence. Purely for testing during initial dev"""
    sentence: str
    generated: datetime
    test: typing.Optional[str] = None
    test2: typing.Union[str, int] = 0
```

Ran code against it to have the schema version saved to the schema registry (under full compatibility). I then altered the model to this, to emulate the changes made under this pull request:

```python
@dataclass
class RandomSentence(faust.Record, AvroModel, serializer='avro_randomsentence'):
    """Randomly Generated Sentence. Purely for testing during initial dev"""
    sentence: str
    generated: datetime
    test: typing.Union[None, str] = None
    test2: typing.Union[int, str] = 0
```

And the change went through and was saved to the schema registry as a new version, as appropriate.

marcosschroh commented:
@xgamer4 amazing! Nice work 👍 . Do you want to update the documentation as well?

xgamer4 commented:
@marcosschroh  Thanks! Documentation has been updated - I added in a bit of explanation to the complex types section on Unions and updated the example
